### DISCUSSION TITLE
Why are my buttons not disabled or rendered like that if command cant be executed

### DISCUSSION THREAD

**Discussion by trashhacker** [AvaloniaUI/Avalonia#17152] [Q&A] [answered]
Buttons have the same apperence even if the command cant be executed (Dark Mode)

**Comment by maxkatz6** [comment]
Button is disabling itself, when Command either is null (can't be reached) or when Command.CanExecute(param) is false.

**Comment by trashhacker** [comment]
yea ok, but its not the case for my buttons ? (:
they are disabled... but not rendered like that

**Reply by stevemonaco** [reply]
This is what you should see in `FluentTheme`'s Dark variant:

![image](https://github.com/user-attachments/assets/0e7e1496-49b5-464c-87ae-8abbf19e35db)

You need to be more specific about your approach and the result you're getting. ie. Xaml/Code, the specific theme you're using, custom style selectors (if applicable), and a screenshot of your result.

**Reply by trashhacker** [reply]
yea like i said (i hope i did) all my buttons and menuitems look like the top button... no matter what

**Reply by stevemonaco** [reply]
Use DevTools and check the actual state of the `Button` at runtime. It should have a `:disabled` pseudoclass and `IsEnabled` property of false. If that's not happening, then your data side is wrong. Not raising `CanExecuteChanged` on the command when dependent properties change, etc.

If it does have those, then your styling is wrong and likely some selector you created.

**Reply by timunie** [reply]
@trashhacker we need to see some sample code, how the hell should we know what you are hacking over there? Try to reproduce in a simple sample and if the issue presist, share the sample with us.

**Reply by trashhacker** [reply]
sorry for my late response, so i have this, its have the empty 
![menuitem](https://github.com/user-attachments/assets/b9016ee1-c862-4c63-b59f-2e66a8cffc49)

im using this for my menuitem itemtemplate

        <DataTemplate x:Key="menuitemcommandtemplate" DataType="{x:Type base:RelayCommand}">
            <MenuItem Command="{Binding}" Header="{Binding Name}">
                <MenuItem.Icon>
                     <Image Source="{Binding Icon, Converter={StaticResource resourceconverter}}" Width="20" Height="20" 
                        VerticalAlignment="Center"/>
                </MenuItem.Icon>
            </MenuItem>
        </DataTemplate>


not styles or themes or something else involved

the CanEexecute method is called correctly and return the correct value
RelayCommand is just a wrapper for ReactiveUI ICommand implementation for Icon and Name properties

**Reply by timunie** [reply]
Command binding is suspicious to me. Are you sure it points to a valid ICommand?

Edit: just saw CanExecute was called so should be correct. However as this is the root element bound to the UI might not recognize the event. Still worth to set the ICommand as a property and bind to that property.

**Reply by trashhacker** [reply]
tried your approach implemeted a porperty for the Command... now all menuitems disabled
check if the CanExecuteChanged Event is fired and result of CanExecute -> working as aspected... 
but still disabled

**Reply by timunie** [reply]
Ahh and you add the MenuItem as DataTemplate. Thats not correct approch here. Instead you need to use a Style or ControlTheme.

**Reply by trashhacker** [reply]
![menuhirarchi](https://github.com/user-attachments/assets/d6abd22b-acee-473e-9ed2-636d6e21f47a)
![menucontent](https://github.com/user-attachments/assets/464fdf57-a445-4194-a465-0c60b0940863)

so the hirarchical tree look like it should?
why is the dropdown empty ?

using this now 

    <ControlTheme x:Key="menuitemcommandtheme" TargetType="MenuItem">
         <Setter Property="Header" Value="{Binding Name}"/>
         <Setter Property="Command" Value="{Binding}"/>
         <Setter Property="Icon">
              <Setter.Value>
                  <DataTemplate>
                      <Image Source="{Binding Icon, Converter={StaticResource resourceconverter}}" Width="16" Height="16" VerticalAlignment="Center"/>
                    </DataTemplate>
                </Setter.Value>
            </Setter>
    </ControlTheme>

**Reply by trashhacker** [reply]
well when i set the ItemContainerTheme its not showing the subitems... even if they exist

**Comment by timunie** [comment]
@trashhacker please provide a very minimal sample. Need to see code to help further

**Comment by trashhacker** [comment]
https://github.com/trashhacker/Avalonia.Repro.Menu

**Comment by trashhacker** [comment]
hmm maybe the problem is general couse i have the same problem in a treeview 
when i set the ItemContainerTheme my treeview is empty :/

**Reply by timunie** [reply, marked as answer]
You're missing BasedOn in the ControlTheme 

`BasedOn="{StaticResource {x:Type MenuItem}}"`

**Reply by trashhacker** [reply]
oh wow, ty a lot
